lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -609,40 +609,44 @@ pub unsafe extern "C" fn read_rdp_output(client_ptr: *mut Client) -> CGOErrCode
 fn read_rdp_output_inner(client: &Client) -> Option<String> {
     let tcp_fd = client.tcp_fd;
     let client_ref = client.go_ref;
+
     // Read incoming events.
     //
     // Wait for some data to be available on the TCP socket FD before consuming it. This prevents
     // us from locking the mutex in Client permanently while no data is available.
     while wait_for_fd(tcp_fd as usize) {
         let mut err = CGOErrCode::ErrCodeSuccess;
-        let res = client
-            .rdp_client
-            .lock()
-            .unwrap()
-            .read(|rdp_event| match rdp_event {
-                RdpEvent::Bitmap(bitmap) => {
-                    let mut cbitmap = match CGOBitmap::try_from(bitmap) {
-                        Ok(cb) => cb,
-                        Err(e) => {
-                            error!(
-                                "failed to convert RDP bitmap to CGO representation: {:?}",
-                                e
-                            );
-                            return;
-                        }
-                    };
-                    unsafe {
-                        err = handle_bitmap(client_ref, &mut cbitmap) as CGOErrCode;
-                    };
-                }
-                // These should never really be sent by the server to us.
-                RdpEvent::Pointer(_) => {
-                    debug!("got unexpected pointer event from RDP server, ignoring");
-                }
-                RdpEvent::Key(_) => {
-                    debug!("got unexpected keyboard event from RDP server, ignoring");
+        let res = client.rdp_client.lock().unwrap().read(|rdp_event| {
+            // This callback can be called multiple times per rdp_client.read()
+            // (if multiple messages were received since the last call). Therefore,
+            // we check that the previous call to handle_bitmap succeeded, so we don't
+            // have a situation where handle_bitmap fails repeatedly and creates a
+            // bunch of repetitive error messages in the logs. If it fails once,
+            // we assume the connection is broken and stop trying to send bitmaps.
```

Comment:
I see that we're not going to repeatedly call handle_bitmap now, but I don't see where we assume the connection is broken and stop trying to send bitmaps. Where is that handled?

Reply:
That's handled the same as it was before:
- `rdp_client.read(cb)` is called (potentially calling the `cb` multiple times)
https://github.com/gravitational/teleport/blob/5012780576ae1d74ca4227a967f42a21360bf0aa/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L619-L649
-  It's result is checked, returning and breaking the `while` loop (which means we stop sending bitmaps) if there was an error
https://github.com/gravitational/teleport/blob/5012780576ae1d74ca4227a967f42a21360bf0aa/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L650-L656
- The callback `cb`'s result is checked, returning and breaking the `while` loop on error
https://github.com/gravitational/teleport/blob/5012780576ae1d74ca4227a967f42a21360bf0aa/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L657-L659